crates/uv-extract/src/stream.rs

```diff
@@ -143,6 +144,16 @@ async fn untar_in(
     mut archive: tokio_tar::Archive<&'_ mut (dyn tokio::io::AsyncRead + Unpin)>,
     dst: &Path,
 ) -> std::io::Result<()> {
+    // Like `tokio-tar`, canonicalize the destination prior to unpacking.
+    let dst = fs_err::tokio::canonicalize(dst).await?;
+
+    // Memoize filesystem calls to canonicalize paths.
+    let mut memo = FxHashSet::default();
+
+    // Delay any directory entries until the end, to ensure that directory permissions do not
+    // interfere with descendant extraction.
+    let mut directories = Vec::new();
```

Comment:
This is related to https://github.com/astral-sh/tokio-tar/pull/2. I forgot that we don't actually use the crate's `unpack` method, we unpack ourselves (since we need to set permissions in a different way).